compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -428,19 +428,57 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                     }
                 }
 
-                err.span_suggestion(
-                    call_span,
-                    &format!("try calling `{}` as a method", ident),
-                    format!("self.{}({})", path_str, args_snippet),
-                    Applicability::MachineApplicable,
-                );
+                if let Some(res) = res {
+                    if let Res::Def(DefKind::Struct, _) = res {
+                        if ns == ValueNS {
+                            // Try Levenshtein algorithm.
+                            let typo_sugg = self.lookup_typo_candidate(path, ns, is_expected, span);
+
+                            // Try context-dependent help if relaxed lookup didn't work.
+                            if self.smart_resolve_context_dependent_help(
+                                &mut err,
+                                span,
+                                source,
+                                res,
+                                &path_str,
+                                &fallback_label,
+                            ) {
+                                // We do this to avoid losing a secondary span when we override the main error span.
+                                self.r.add_typo_suggestion(&mut err, typo_sugg, ident_span);
+                                return (err, candidates);
+                            }
+                        } else {
+                            err.span_suggestion(
+                                call_span,
+                                &format!("try calling `{}` as a method", ident),
+                                format!("self.{}({})", path_str, args_snippet),
+                                Applicability::MachineApplicable,
+                            );
+                        }
+                    } else {
+                        err.span_suggestion(
+                            call_span,
+                            &format!("try calling `{}` as a method", ident),
+                            format!("self.{}({})", path_str, args_snippet),
+                            Applicability::MachineApplicable,
+                        );
+                    }
+                } else {
+                    err.span_suggestion(
+                        call_span,
+                        &format!("try calling `{}` as a method", ident),
+                        format!("self.{}({})", path_str, args_snippet),
+                        Applicability::MachineApplicable,
+                    );
+                }
+
                 return (err, candidates);
             }
         }
 
         // Try Levenshtein algorithm.
         let typo_sugg = self.lookup_typo_candidate(path, ns, is_expected, span);
-        // Try context-dependent help if relaxed lookup didn't work.
+        // Do the same as above. These two heuristics *link* to each other and use the same condition.
```

Comment:
`smart_resolve_context_dependent_help` is a huge function, this comment applies only to a tiny part of it, could you move it to the relevant code?